src/jvm/main/org/jetbrains/kotlinx/lincheck/LinChecker.kt

```diff
@@ -50,80 +52,65 @@ class LinChecker (private val testClass: Class<*>, options: Options<*, *>?) {
         check(testConfigurations.isNotEmpty()) { "No Lincheck test configuration to run" }
         for (testCfg in testConfigurations) {
             val failure = testCfg.checkImpl()
-            if (failure != null) return failure
+            if (failure != null)
+                return failure
         }
         return null
     }
 
     private fun CTestConfiguration.checkImpl(): LincheckFailure? {
-        val exGen = createExecutionGenerator(testStructure.randomProvider)
-        for (i in customScenarios.indices) {
-            val verifier = createVerifier()
-            val scenario = customScenarios[i]
-            scenario.validate()
-            reporter.logIteration(i + 1, customScenarios.size, scenario)
-            val failure = scenario.run(this, verifier)
-            if (failure != null) return failure
-        }
         var verifier = createVerifier()
-        repeat(iterations) { i ->
+        val generator = createExecutionGenerator(testStructure.randomProvider)
+        val randomScenarios = generateSequence {
```

Comment:
I don't think you need a sequence here, as right after that you generate all the scenarios.

Reply:
With sequences I wanted to preserve the same logic as before, so that the scenarios are generated lazily on demand. But also I wanted to have single loop to handle both custom and random scenarios. 

So I join into a signle sequence:
* custom scenarios list (converting it into a sequence) and 
* a sequence generating random scenarios -- I truncate it using `take` method, so that only `N = number of iterations` scenarios are generated. 

Please note, that `Sequence.take` [method](https://github.com/JetBrains/kotlin/blob/0938b46726b9c6938df309098316ce741815bb55/libraries/stdlib/common/src/generated/_Sequences.kt#L556) does not forces the evaluation of the sequence. 
So the random scenarios are generated lazily on demand.
If we discover failure earlier, subsequent scenarios will not be generated.

Reply:
I can add a comment to the code to clarify this logic.

Reply:
That would be wonderful!